File: LeetCode 0200 Number of Islands.py
class Solution(object): # 95%, 16%
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid: return 0
        ans = 0
        xlen, ylen = len(grid), len(grid[0])
        
        def mark(x, y): # recursion，记录当前岛块已visited，并记录其上下左右
            if grid[x][y] == "1": # 保证不是“0”非岛屿或者“2”已被visited
                grid[x][y] = "2" # 记录方式为标记为“2”
                if x >= 1: mark(x - 1, y) # 左
                if y >= 1: mark(x, y - 1) # 上
                if x < xlen - 1: mark(x + 1, y) # 右
                if y < ylen - 1: mark(x, y + 1) # 下
        
        for x in range(xlen): # main，逐个排查
            for y in range(ylen):
                if grid[x][y] == "1": # 排查到未visit过的岛屿
                    ans += 1
                    mark(x, y) # 标记，思路类似于union find
        return ans

File: LeetCode 0200 Number of Islands.java
class Solution {
    public int numIslands(char[][] grid) { // 99.9, 65
        if (grid == null || grid.length == 0 || grid[0].length == 0)
            return 0;
        int r = grid.length;
        int c = grid[0].length;
        int count = 0;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                if (grid[i][j] == '1') {
                    mark(grid, i, j);
                    count++;
                }
            }
        }
        return count;
    }
    
    private void mark(char[][]grid, int i, int j) {
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length)
            return;
        if (grid[i][j] != '1')
            return;
        grid[i][j] = '2';
        mark(grid, i + 1, j);
        mark(grid, i - 1, j);
        mark(grid, i, j + 1);
        mark(grid, i, j - 1);
    }
}